src/shims/windows/sync.rs

```diff
@@ -0,0 +1,161 @@
+use rustc_target::abi::Size;
+
+use crate::*;
+
+// Locks are pointer-sized pieces of data, initialized to 0.
+// We use them to count readers, with usize::MAX representing the write-locked state.
+
+fn deref_lock<'mir, 'tcx: 'mir>(
+    ecx: &mut MiriEvalContext<'mir, 'tcx>,
+    lock_op: OpTy<'tcx, Tag>,
+) -> InterpResult<'tcx, MPlaceTy<'tcx, Tag>> {
+    // `lock` is a pointer to `void*`; cast it to a pointer to `usize`.
+    let lock = ecx.deref_operand(lock_op)?;
+    let usize = ecx.machine.layouts.usize;
+    assert_eq!(lock.layout.size, usize.size);
+    Ok(lock.offset(Size::ZERO, MemPlaceMeta::None, usize, ecx)?)
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    #[allow(non_snake_case)]
+    fn AcquireSRWLockExclusive(
```

Comment:
This function could just invoke its `Try*` equivalent and cause the deadlock error in case the result is `0`. That should help deduplicating the logic

Reply:
I think you looked at an old intermediate commit -- this changed quite a bit for the final PR.

Reply:
I entirely re-implemented rwlocks later and figured I would keep the history. Sorry for the confusion.

Reply:
:laughing: I didn't get that far yet

Reply:
so the thread id assertions are gone because the logic would now support threading? (I know we don't support thread starting yet at all, but the lock logic works now correctly for threads?)

Reply:
Correct. It's the same logic as what we use on POSIX targets.